astropy/utils/parsing.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Wrappers for PLY to provide thread safety.
+"""
+
+import contextlib
+import functools
+import re
+import os
+import threading
+
+
+TAB_HEADER = """# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# This file was automatically generated from ply. To re-generate this file,
+# remove it from this folder, then build astropy and run the tests in-place:
+#
+#   python setup.py build_ext --inplace
+#   pytest {package}
+#
+# You can then commit the changes to this file.
+
+"""
+_LOCK = threading.RLock()
+
+
+def _add_tab_header(filename, package):
+    with open(filename, 'r') as f:
+        contents = f.read()
+
+    with open(filename, 'w') as f:
+        f.write(TAB_HEADER.format(package=package))
+        f.write(contents)
+
+
+@contextlib.contextmanager
+def _patch_get_caller_module_dict(module):
+    """Temporarily replace the module's get_caller_module_dict.
+
+    This is a function inside ``ply.lex`` and ``ply.yacc`` (each has a copy)
+    that is used to retrieve the caller's local symbols. Here, we patch the
+    function to instead retrieve the grandparent's local symbols to account
+    for a wrapper layer.
+    """
+    original = module.get_caller_module_dict
+
+    @functools.wraps(original)
+    def wrapper(levels):
+        # Add 2, not 1, because the wrapper itself adds another level
+        return original(levels + 2)
+
+    module.get_caller_module_dict = wrapper
+    yield
+    module.get_caller_module_dict = original
+
+
+def lex(lextab, package, reflags=int(re.VERBOSE)):
+    """Create a lexer from local variables.
+
+    It automatically compiles the lexer in optimized mode, writing to
```

Comment:
please reformat to a numpydoc type docstring for the parameters rather than have narrative only.

Reply:
Done